### PR TITLE
copr: bit column support push down to TiKV

### LINKED ISSUE #30738: The predicate on the `bit` column cannot be pushed down
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
create table t(a int, b int, c bit(1), key(a, b, c));
explain select * from t where c=0 order by a, b limit 10;
```

### 2. What did you expect to see? (Required)
```
+------------------------------+---------+-----------+---------------------------+-------------------------------+
| id                           | estRows | task      | access object             | operator info                 |
+------------------------------+---------+-----------+---------------------------+-------------------------------+
| Limit_12                     | 10.00   | root      |                           | offset:0, count:10            |
| └─IndexReader_30             | 10.00   | root      |                           | index:Limit_29                |
|   └─Limit_29                 | 10.00   | cop[tikv] |                           | offset:0, count:10            |
|     └─Selection_28           | 10.00   | cop[tikv] |                           | gt(test.t.c, 0)               |
|       └─IndexFullScan_27     | 30.00   | cop[tikv] | table:t, index:a(a, b, c) | keep order:true, stats:pseudo |
+------------------------------+---------+-----------+---------------------------+-------------------------------+
```

### 3. What did you see instead (Required)
```
+----------------------------+---------+-----------+----------------------------+-------------------------------+
| id                         | estRows | task      | access object              | operator info                 |
+----------------------------+---------+-----------+----------------------------+-------------------------------+
| Limit_14                   | 10.00   | root      |                            | offset:0, count:10            |
| └─Selection_21             | 10.00   | root      |                            | eq(test.t.c, 0)              |
|   └─IndexReader_23         | 10.00   | root      |                            | index:IndexFullScan_22        |
|     └─IndexFullScan_22     | 10.00   | cop[tikv] | table:t, index:a(a, b, c) | keep order:true, stats:pseudo |
+----------------------------+---------+-----------+----------------------------+-------------------------------+
```

### 4. What is your TiDB version? (Required)

Release Version: v5.4.0-alpha-411-g04a9618f5
Edition: Community
Git Commit Hash: 04a9618f5cccd2ef262b2ef04aeab1a6770c862a
Git Branch: HEAD
UTC Build Time: 2021-12-15 07:14:59
GoVersion: go1.17.3
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

### PR DESCRIPTION
**Pull request by xiejiandong**
### What problem does this PR solve?
Issue Number: close #30738

Problem Summary:
Old TiDB version forbit Bit column push down;  Because TiKV not support it.
When now TiKV support Bit column push down， The PR to fit it. 

Solve Methods：
TiKV PR:  https://github.com/tikv/tikv/pull/11968
TiDB:
1、After TiKV PR,  TiDB Open TypeBit to push down； 
2、TiDB pass ret_field_type.Flen to Tikv，hint bit column len,
      To fit cast、ascii、binary etc scallar functions；

Tests
- [x] Unit test
- [x] [Integration test](https://github.com/tikv/copr-test)
- [x] [Manual test](https://docs.google.com/document/d/1jPOzbBZlxkdusJPQfh38IW7KfDv08YERKqpMAfG7oEM/edit#)

Side effects
- None

### Release note

```release-note
Added support for Bit column push down. [#30738]
```

### CODE REVIEW

File: expression/expression.go
Comment (Yisaer) [inline expression/expression.go:1257]: Is `tiflash` support `bit` column yet?

**Review by Yisaer — COMMENTED** [review]
we shall add the intergration test and wait https://github.com/tikv/tikv/pull/11894 be merged.

File: expression/builtin_cast.go
Comment (Yisaer) [inline expression/builtin_cast.go:1957]:
```suggestion
		// For TypeBit, castAsString will make length as int(( bit_len + 7 ) / 8) bytes due to
		// TiKV needs the bit's real len during calculating, eg: ascii(bit).
```

File: planner/core/expression_rewriter.go
Comment (Yisaer) [inline planner/core/expression_rewriter.go:1136]:
```suggestion
			// see detailed comment in `WrapWithCastAsString`
```

File: expression/expression.go
Comment (Yisaer) [reply expression/expression.go:1257]: As we haven't supported push bit into tiflash, we should keep `Typebit` here

**Review by Yisaer — COMMENTED** [review]
rest LGTM

**Review by Yisaer — COMMENTED** [review]
Also, I think we need to add a test case in `TestScalarFunctionPushDown` with bit column

File: expression/expression.go
Comment (xiejiandong) [reply expression/expression.go:1257]: ok

File: expression/builtin_cast.go
Comment (xiejiandong) [reply expression/builtin_cast.go:1957]: ok

File: planner/core/expression_rewriter.go
Comment (xiejiandong) [reply planner/core/expression_rewriter.go:1136]: ok

**Comment by Yisaer** [comment]
/hold cancel

due to bit push down has already finished in tikv

**Review by Yisaer — APPROVED** [review]
lgtm

File: planner/core/expression_rewriter.go
Comment (winoros) [inline planner/core/expression_rewriter.go:1137]: Why we still need to add this if block after we added the same codes in the `WrapWithCastAsString`?

**Comment by Yisaer** [comment]
/hold

due to I use this pr and latest tikv and test following sql and get error:

```sql
drop table if exists  t1,t2;
create table t1(a bit(10), b int);
create table t2(a bit(10), b int);
insert into t1 values  ('3', 3);
insert into t2 values ('2', 2) ;
explain select b from t1 where t1.b > (select min(t2.b) from t2 where t2.a < t1.a);
select b from t1 where t1.b > (select min(t2.b) from t2 where t2.a < t1.a);

```

File: expression/builtin_cast.go
Comment (Yisaer) [inline expression/builtin_cast.go:1965]: Is there any testcase to cover this? eg: push ascii bit column down and assert the result.

File: planner/core/expression_rewriter.go
Comment (xiejiandong) [reply planner/core/expression_rewriter.go:1137]: Function ascii()、concat() will call through expression/builtin_cast.go::WrapWithCastAsString；
Function cast()、binary()  will call through planner/core/expression_rewriter.go::Leave;
The two kinds of functions be called not the same way,  so there are two positions needed to be updated.
And I don't know is there any other position need to update. So by copr-test, I only updated the 2 positions and passed the test;

File: expression/builtin_cast.go
Comment (xiejiandong) [reply expression/builtin_cast.go:1965]: Ok, I will add it.

**Comment by Yisaer** [comment]
duplicated with https://github.com/pingcap/tidb/pull/32497

**Pull request closed without merging**
